moveos/moveos-verifier/src/verifier.rs

```diff
@@ -40,68 +42,49 @@ where
                     fdef,
                     module.self_id(),
                 ));
-            } else if fdef.is_entry {
+            }
+
+            if fdef.is_entry {
                 return Err(vm_error_for_init_func_checking(
                     StatusCode::INVALID_MAIN_FUNCTION_SIGNATURE,
                     "init function should not entry function",
                     fdef,
                     module.self_id(),
                 ));
-            } else {
-                let function_id =
-                    FunctionId::new(module.self_id(), INIT_FN_NAME_IDENTIFIER.clone());
-                let loaded_function = session.load_function(
-                    &module.self_id(),
-                    &function_id.function_name,
-                    vec![].as_slice(),
-                )?;
-                let parameters_usize = loaded_function.parameters.len();
-                if parameters_usize != 1 && parameters_usize != 2 {
+            }
+
+            let view = BinaryIndexedView::Module(module);
+            let func_parameter_signatures = view.signature_at(fhandle.parameters);
+            let func_parameter_vec = &func_parameter_signatures.0;
+
+            if func_parameter_vec.len() != 1 && func_parameter_vec.len() != 2 {
+                return Err(vm_error_for_init_func_checking(
+                    StatusCode::NUMBER_OF_TYPE_ARGUMENTS_MISMATCH,
+                    "init function only should have two parameter with signer or storageContext",
+                    fdef,
+                    module.self_id(),
+                ));
+            }
+
+            for st in func_parameter_vec {
+                let (is_allowed, struct_name_opt) = is_allowed_init_func_param(&view, st);
```

Comment:
maybe `is_allowed_init_func_param` can be combined with the above function to verify the number of parameters,  considering the same error is thrown?